### ISSUE TITLE
Terraform crash when using count on a list of objects

### ISSUE THREAD

**Issue by JamesDLD** [hashicorp/terraform#22082]
### Terraform Version

```
Terraform v0.12.3
+ provider.azurerm v1.31.0
```

### Terraform Configuration Files

```hcl
#Variable 
virtual_networks = [
  {
    id            = "1"
    address_space = ["10.0.0.0/16"]
    subnets = [
      {
        name           = "test1"
        address_prefix = "10.0.0.0/24"
      },
      {
        name                     = "test2"
        address_prefix           = "10.0.1.0/24"
        security_group_iteration = "0"
      },
      {
        name           = "AzureBastionSubnet"
        address_prefix = "10.0.2.0/27"
      },
    ]
  },
  {
    id            = "2"
    address_space = ["10.1.0.0/16"]
    subnets = [
      {
        name                     = "test1"
        address_prefix           = "10.1.0.0/24"
        security_group_iteration = "0"
      },
    ]
  },
]

#Resource 
resource "azurerm_virtual_network" "vnets" {
  count               = length(var.virtual_networks)
  name                = "${var.product["entity_suffix"]}-${var.product["product_name"]}-${data.azurerm_resource_group.network.location}-${var.product["env"]}-vnet${var.virtual_networks[count.index]["id"]}"
  location            = data.azurerm_resource_group.network.location
  resource_group_name = data.azurerm_resource_group.network.name
  address_space       = var.virtual_networks[count.index]["address_space"]

  dynamic "subnet" {
    for_each = var.virtual_networks[count.index]["subnets"]
    content {
      name           = lookup(subnet.value, "name", null)
      address_prefix = lookup(subnet.value, "address_prefix", null)
      security_group = lookup(subnet.value, "security_group_iteration", null) == null ? null : element(azurerm_network_security_group.nsgs.*.id, lookup(subnet.value, "security_group_iteration"))
    }
  }

  tags = data.azurerm_resource_group.network.tags
}
```

### Debug Output

https://github.com/JamesDLD/debug/blob/master/debug.log

### Crash Output

https://github.com/JamesDLD/debug/blob/master/crash.log

### Expected Behavior

Create 2 vnet. No error when removing the second vnet.

### Actual Behavior

```hcl
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce

https://github.com/JamesDLD/debug

**Comment by JamesDLD** [comment]
I just founded where does come the crash, it's linked to the way I declared my variable.
Is this a bug or did I do a mistake with this declaration?

This prevent from having a crash -->
```hcl
variable "virtual_networks" {
  description = "The virtual networks with their properties."
  type=any
}
```

This implies a crash -->
```hcl
variable "virtual_networks" {
  description = "The virtual networks with their properties."
  type = list(object({
    id            = string
    address_space = list(string)
    subnets       = any
  }))
}
```

**Comment by JamesDLD** [comment]
Same issue with Terraform v0.12.5.

**Comment by danieldreier** [comment]
@JamesDLD I want to apologize for how long it's taken to respond to this crash report. Thank you for reporting this - genuinely appreciate it.

I tested this against 0.12.26, and was able to reproduce the problem. I tested against 0.13.0 beta 1, and was no longer able to reproduce it. The test case I used is in https://github.com/danieldreier/terraform-issue-reproductions/tree/master/22082 in case someone you want to try.

I'm going to close this as resolved because I think that this was fixed in 0.13.0. If I'm testing wrong and you're still able to reproduce it, please let me know at ddreier@hashicorp.com and I'm happy to re-open this.
